proposals/type-classes.md

```diff
@@ -0,0 +1,298 @@
+Furthermore introduction of type classes improves usages for `reified` generic functions with a more robust approach that does not require those to be `inline` or `reified`.
+
+## Motivation
+
+* Support Type class evidence compile time verification.
+* Support a broader range of Typed FP patterns.
+* Enable multiple extension functions groups for type declarations.
+* Enable compile time DI through the use of the Type class pattern.
+* Enable better compile reified generics without the need for explicit inlining.
+* Enable definition of polymorphic functions whose constrains can be verified at compile time in call sites.
+
+## Description
+
+We propose to use the existing `interface` semantics allowing for generic definition of type classes and their instances with the same style interfaces are defined
+
+```kotlin
+interface Monoid<A> {
+    fun A.combine(b: A): A
+    fun empty(): A
+}
+```
+
+The above declaration can serve as target for implementations for any arbitrary data type.
+In the implementation below we provide evidence that there is a `Monoid<Int>` instance that enables `combine` and `empty` on `Int`
+
+```kotlin
+package intext
+
+object IntMonoid : Monoid<Int> {
+    fun Int.combine(b: Int): Int = this + b
+    fun empty(): Int = 0
+}
+```
+
+```
+import intext.IntMonoid
+
+1.combine(2) // 3
+Int.empty() // 0
+```
+
+Because of this constrain where we are stating that there is a `Monoid` constrain for a given type `A` we can also encode polymorphic definitions based on those constrains:
+
+```kotlin
+import intext.IntMonoid
+
+fun <A> add(a: A, b: A): A given Monoid<A> = a.combine(b)
+add(1, 1) // compiles
+add("a", "b") // does not compile: No `String: Monoid` instance defined in scope
+```
+
+## Compile Time Dependency Injection
+
+On top of the value this brings to typed FP in Kotlin it also helps in OOP contexts where dependencies can be provided at compile time:
+
+```kotlin
+interface Context<A> {
+  fun A.config(): Config
+}
+```
+
+```kotlin
+package prod
+
+object ProdContext: Context<Service> {
+  fun Service.config(): Config = ProdConfig
+}
+```
+
+```kotlin
+package test
+
+object TestContext: Context<Service> {
+  fun Service.config(): Config = TestConfig
+}
+```
+
+```kotlin
+package prod
+
+service.config() // ProdConfig
+```
+
+```kotlin
+package test
+
+service.config() // TestConfig
+```
+
+## Overcoming `inline` + `reified` limitations
+
+Type classes allow us to workaround `inline` `reified` generics and their limitations and express those as type classes instead:
+
+```kotlin
+interface Reified<A> {
+    val selfClass: KClass<A>
+}
+```
+
+Now a function that was doing something like:
+
+```kotlin
+inline fun <reified A> foo() { .... A::class ... }
+```
+
+can be replaced with:
+
+```kotlin
+fun <A> fooTC(): Klass<A> given Reified<A> { .... A.selfClass ... }
```

Comment:
This was just a particular example of cases where you know in a inline reified context the concrete type params you want to support and then you can provide instances of `Reified` for those but it does not replace inline reified generics instrospected at runtime. that section originated from this comment https://github.com/Kotlin/KEEP/pull/87#issuecomment-333915196

Reply:
Still, it's not clear from the proposal, how the instances are constructed, and the text suggests that it's ratehr a repacement than a complement. I think it needs to be clarified

Reply:
In Scala the compiler injects an instance in the actual argument default value because they use implicit arguments but in Kotlin given this was related to extensions the compiler will enable the extension syntax provided by the typeclass in the same way extensions functions work when brought into scope. At the end of the day the compiler needs to inject bytecode that creates a new object if it's a class or uses the singleton value if an object similar to how `with(instance) { }` works now. Open to ideas as to how this should better work :)

Reply:
There's a load of open questions:
- where is the `Reified` interface defined?
- what's its exact API?
- how many instances need being created and how they are cached?
- do new classes need to be created to instantiate `Reified` (a positive answer to this sounds like a show stopper to me)?

Reply:
> There's a load of open questions:
> where is the Reified interface defined?
> what's its exact API?
I would not use `Reified` as a type class but was an example provided by @elizarov as having similar semantics.
A type class interface may be defined anywhere in user or library code such as 
```kotlin
package x
extension interface Reified<A> {
  fun A.kclass(): KClass<A>
}
```
```kotlin
package y
extension object StringReified : Reified<String> {
  fun String.kclass(): KClass<A> = String::class
}
```
```kotlin
package z

import y.*

fun <A> getKClass(): KClass<A> given Reified<A> = A.kclass()

object test {
  val sc = getKClass<String>() //compiles
  val ic = getKclass<Int>() // does not compile
}
```
> how many instances need being created and how they are cached?
In the case of using an `object` the instance itself is the value. In the case of defining instances that declare further evidence of other type classes in terms of a `class`:
```kotlin
extension class OptionMonoid<A>: Monoid<Option<A>> given Monoid<A>
```
The compiler may need to instantiate it if we allow for mutable state or just cache the instance.
> do new classes need to be created to instantiate Reified (a positive answer to this sounds like a show stopper to me)?
I don't think so because the user declared instances themselves are the evidence. We restricted in the proposal usage to just `object` and `class` but allowing `val` or `fun` providers is also an option for resolution, though that looks a lot more than Scala implicits and the purpose is that injection and resolution would be restricted to type classes and all of those are parametric to at least one type param.

Reply:
> `Reified`

This suggests that the instance has to be defined manually every time: an instance for String, instance for Int, and so on. This can't be practical, TBH. And if you expect the compiler to automatically define these, there has to be a caching mechanism or something described in this proposal

Reply:
> The compiler may need to instantiate it if we allow for mutable state or just cache the instance.

There's no way in the language to guarantee purity/immutability, so the compiler has to assume mutable state most of the time

Reply:
> We restricted in the proposal usage to just object and class but allowing val or fun providers is also an option for resolution

In terms of resolution, there's not much of a difference between `object` and `val`

Reply:
`Reified` is not a good example for type classes when the use case is instrospection of reified runtime generics with reflection. 
Coupling behavior to data types without inheritance that are extensible by users ad-hoc where needed is the killer feature. 
Perhaps I should remove the `Reified` example from the proposal so it does not lead to confusion as it is not a real replacement when you want to do runtime reflection generalized to all types.